sys/sched.c

```diff
@@ -190,6 +190,13 @@ void preempt_enable(void) {
   thread_t *td = thread_self();
   assert(td->td_pdnest > 0);
   td->td_pdnest--;
+
+  if (td->td_pdnest == 0 && td->td_flags & TDF_NEEDSWITCH) {
```

Comment:
If `td->td_pdnest == 0` and `td->td_flags & TDF_NEEDSWITCH` are true, we enter the *if statement* and get preempted by interrupt, and after return we yield though `TDF_NEEDSWITCH` is not marked anymore. Isn't that a race condition?

Reply:
Yes. I think so. I've moved `if` into spinlock.
BTW At first glance in FreeBSD they have similar situation if an interrupt happened between lines 209 and 210. http://bxr.su/FreeBSD/sys/kern/kern_switch.c#198

Reply:
@cahirwpz as I said in FreeBSD they had similar issue. I thought I would fix it, so I would have second commit in FreeBSD and therefore get more fame and glory. But someone 28 days ago noticed it as well and forestalled me... :disappointed: 
https://github.com/freebsd/freebsd/commit/e4785b15b237e7ed6ace7ace14cb2acb1287af29